gnovm/pkg/gnolang/type_check.go

```diff
@@ -215,6 +215,141 @@ func assertAssignableTo(xt, dt Type, autoNative bool) {
 	}
 }
 
+func assertValidConstExpr(store Store, last BlockNode, n *ValueDecl, expr Expr) {
+	if n.Type != nil {
+		nt := evalStaticType(store, last, n.Type)
+		if xnt, ok := nt.(*NativeType); ok {
+			nt = go2GnoBaseType(xnt.Type)
+		}
+
+		if _, ok := baseOf(nt).(PrimitiveType); !ok {
+			panic(fmt.Sprintf("invalid constant type %s", nt.String()))
+		}
+	}
+
+	assertValidConstantExprRecursively(store, last, expr, nil)
+}
+
+func assertValidConstantExprRecursively(store Store, last BlockNode, currExpr, parentExpr Expr) {
```

Comment:
```suggestion
func assertValidConstValue(store Store, last BlockNode, currExpr, parentExpr Expr) {
```

Reply:
should be done here https://github.com/gnolang/gno/pull/2828/commits/54878f00b8eb36747d884e86cd5df86f7be76f1b